bigtable_input.py

```diff
@@ -45,22 +47,66 @@ def get_game_range_row_names(game_begin, game_end):
     Sample row name:
       g_0000000001_m001
 
-    To capture the range of all moves in the two given games,
-    the end row will need to go up to g_00..(N+1).
-
     Args:
       game_begin:  an integer of the beginning game number.
-      game_end:  an integer of the ending game number, inclusive.
+      game_end:  an integer of the ending game number, exclusive.
 
     Returns:
       The two string row numbers to pass to Bigtable as the row range.
     """
     row_fmt = 'g_{:0>10}_'
-    return row_fmt.format(game_begin), row_fmt.format(game_end + 1)
+    return row_fmt.format(game_begin), row_fmt.format(game_end)
+
+
+def make_single_array(ds, batch_size=8*1024):
+    full_result = None
+    with tf.Session() as sess:
+        ds = ds.batch(batch_size)
+        iterator = ds.make_initializable_iterator()
+        sess.run(iterator.initializer)
+        get_next = iterator.get_next()
+        try:
+            while True:
+                result = sess.run(get_next)
+                if full_result is None:
+                    full_result = result
+                else:
+                    full_result = np.concatenate((full_result, result))
+        except tf.errors.OutOfRangeError:
+            pass
+    return full_result
+
+
+def get_moves_from_games(start_game, end_game, moves, shuffle,
+                         column_family, column):
+    start_row, end_row = get_game_range_row_names(start_game, end_game)
+    # NOTE:  Choose a probability high enough to guarantee at least the
+    # required number of moves, by using a slightly lower estimate
+    # of the total moves, then trimming the result.
+    total_moves = count_moves_in_game_range(start_game, end_game)
+    probability = moves / (total_moves * 0.99)
+    utils.dbg('Row range: %s - %s; total moves: %d; probability %.3f' % (
+        start_row, end_row, total_moves, probability))
+    shards = 8
+    ds = _tf_table.parallel_scan_range(start_row, end_row,
+                                       probability=probability,
+                                       num_parallel_scans=shards,
+                                       columns=[(column_family, column)])
+    if shuffle:
+        rds = tf.data.Dataset.from_tensor_slices(
+            tf.random_shuffle(tf.range(0, shards, dtype=tf.int64)))
```

Comment:
I don't think shuffling the shard orders does anything if you're reading them all at the same time anyway?

Reply:
Good point. When I was testing, I alternated between `Dataset.interleave` and `tf.contrib.data.parallel_interleave`, so it was important to shuffle the shards. And even in this case, although I believe you're right that the parallel interleave will likely not be ordered, it doesn't cost anything to intentionally disorder it here. (After all, `parallel_scan_range` itself doesn't promise ordered results, and they're not, but they're not disordered _enough_, hence this update.)